connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -1425,6 +1426,256 @@ public WorkerTask doBuild(Task task,
+        // but we know that that topic is the same as the worker-global offsets topic, we ignore the worker-global
+        // offset store and build a store backed exclusively by a connector-specific offsets store.
+        // We cannot under any circumstances build an offset store backed exclusively by the worker-global offset store
+        // as that would prevent us from being able to write source records and source offset information for the task
+        // with the same producer, and therefore, in the same transaction.
+        if (sameOffsetTopicAsWorker(connectorOffsetsTopic, producerProps)) {
+            return ConnectorOffsetBackingStore.withOnlyConnectorStore(
+                    () -> LoggingContext.forTask(id),
+                    connectorStore,
+                    connectorOffsetsTopic,
+                    topicAdmin
+            );
+        } else {
+            return ConnectorOffsetBackingStore.withConnectorAndWorkerStores(
+                    () -> LoggingContext.forTask(id),
+                    globalOffsetBackingStore,
+                    connectorStore,
+                    connectorOffsetsTopic,
+                    topicAdmin
+            );
+        }
+    }
+
+    /**
+     * Gives a best-effort guess for whether the given offsets topic is the same topic as the worker-global offsets topic.
+     * Even if the name of the topic is the same as the name of the worker's offsets topic, the two may still be different topics
+     * if the connector is configured to produce to a different Kafka cluster than the one that hosts the worker's offsets topic.
+     * @param offsetsTopic the name of the offsets topic for the connector
+     * @param producerProps the producer configuration for the connector
+     * @return whether it appears that the connector's offsets topic is the same topic as the worker-global offsets topic.
+     * If {@code true}, it is guaranteed that the two are the same;
+     * if {@code false}, it is likely but not guaranteed that the two are not the same
+     */
+    private boolean sameOffsetTopicAsWorker(String offsetsTopic, Map<String, Object> producerProps) {
+        // We can check the offset topic name and the Kafka cluster's bootstrap servers,
+        // although this isn't exact and can lead to some false negatives if the user
+        // provides an overridden bootstrap servers value for their producer that is different than
+        // the worker's but still resolves to the same Kafka cluster used by the worker.
+        // At the moment this is probably adequate, especially since we don't want to put
+        // a network ping to a remote Kafka cluster inside the herder's tick thread (which is where this
+        // logic takes place right now) in case that takes a while.
+        return offsetsTopic.equals(config.offsetsTopic())
+                && config.bootstrapServers().equals(producerProps.get(BOOTSTRAP_SERVERS_CONFIG));
```

Comment:
True, we can make this order-independent by comparing sets. Not too much work, so I've done that.

I have a long-term vision of relocating this logic into the `WorkerTask` class (or its subclasses) so that we can ping the remote Kafka cluster, grab its ID there, and decide at that point whether or not we need to construct an additional offsets store, all without having to worry about blocking up the herder tick thread. This would obviate the need for users to do any additional work at all, but it would require a significant refactoring of this part of the code base, so I've held off on it for now.